python_modules/dagster/dagster/_core/definitions/freshness_policy.py

```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Mapping, Optional
+
+import pendulum
+from croniter import croniter
+
+from dagster._annotations import experimental
+
+from .events import AssetKey
+
+
+@experimental
+class FreshnessPolicy(ABC):
+    """A FreshnessPolicy is a policy that defines how up-to-date a given asset is expected to be.
+    We calculate the current time of the data within an asset by traversing the history of asset
+    materializations of upstream assets which occured before the most recent materialization.
+
+    This gives a lower bound on the most recent records that could possibly be incorporated into the
+    current state of the asset to which this policy is attached.
+    """
+
+    @abstractmethod
+    def minutes_late(
+        self,
+        evaluation_time: datetime,
+        upstream_materialization_times: Mapping[AssetKey, Optional[datetime]],
+    ) -> Optional[float]:
```

Comment:
In what case would `None` get returned?

Reply:
None gets returned if the data time for any of its upstream keys is None. In these cases, you know something's wrong (you have no record of the materialization of your upstream asset which was used to create this one), but the amount wrong that you are is unknowable.